### PR TITLE
Tests for RC

### PR DESCRIPTION
**Pull request by laktek**
## What kind of change does this PR introduce?

Tests for client library (based on #11 - rebased & updated against `rc` branch)

## What is the current behavior?

No tests for `rc` branch

## What is the new behavior?

Some amount of tests for the client: invoke, auth, headers, body params.
Core components to build tests like reporting and relay container start.

## Additional context

Tests are based on jest, test-containers-node. I've removed the allure coverage reporting which was in #11 PR as there were some issues in getting it to work with jest 28. We can add coverage reporting in a separate PR.

### CODE REVIEW

**Review by inian — APPROVED** [review]

**Comment by inian** [comment]
FormData needs to be polyfilled for running tests

**Comment by laktek** [comment]
> FormData needs to be polyfilled for running tests

Not needed if tests runs in Node 18+ (which will be LTS from Oct 2022), We'll make a note about it for running tests in contributing docs. Also, `node-formdata` polyfill is kinda outdated and not actively maintained (don't think it's a good idea to bring it in as a dep)

**Comment by egor-romanov** [comment]
Should we run tests workflow on prs to rc?
Other than that 👍

Also, there is a hijack connection test. Should we return an error if a customer tries to hijack in function, or should we allow that?

File: test/utils/binaries.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+ * @returns The array buffer that contains the string.
+ */
+export function str2ab(str: string) {
+  var buf = new ArrayBuffer(str.length * 2) // 2 bytes for each char
+  var bufView = new Uint16Array(buf)
+  for (var i = 0, strLen = str.length; i < strLen; i++) {
+    bufView[i] = str.charCodeAt(i)
+  }
+  return buf
+}
```
Comment (soedirgo) [inline test/utils/binaries.ts:27]: Better alternative: `new TextEncoder().encode('mystring')`

Also, `ab2str` is unused

**Review by soedirgo — COMMENTED** [review]
A couple nits, but otherwise looks good :+1:

**Comment by laktek** [comment]
> Should we run tests workflow on prs to rc?
> Other than that 👍

@egor-romanov I think so! will make a separate PR for GitHub actions.  

> Also, there is a hijack connection test. Should we return an error if a customer tries to hijack in function, or should we allow that?

It currently returns an error when a user tries to hijack in function. But not sure if the error is descriptive, we can create a separate ticket to add better error handling for this instance.

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+name: Tests
+
+# Controls when the workflow will run
+on:
+  # Triggers the workflow on push or pull request events but only for the main branch
+  push:
+    branches: [ main ]
+  pull_request:
+    branches: [ main ]
+
```
[the remark this answers is not in the document]
Comment (laktek) [reply .github/workflows/tests.yml:11]: @soedirgo If you don't mind, shall I make a separate PR for GitHub actions and merge it first?

File: package.json [hunk 3]
```diff
@@ -35,15 +35,27 @@
   },
   "homepage": "https://github.com/supabase/functions-js#readme",
   "dependencies": {
-    "cross-fetch": "^3.1.5"
+    "cross-fetch": "^3.1.5",
+    "form-data": "^4.0.0"
```
[the remark this answers is not in the document]
Comment (laktek) [reply package.json]: good catch! fixed in 640fc51

**Pull request merged**